File: 2022/D09/d09.js
/*
Advent Of Code 2022
Day 9: Rope Bridge 1 & 2

https://adventofcode.com/2022/day/9
*/

const fs = require('fs');
const path = require('path');

const YEAR = '2022';
const DAY = '09';

function readInput(filename) {
  return fs
    .readFileSync(path.join(`${YEAR}`, `D${DAY}`, filename))
    .toString()
    .split(`\n`)
    .filter((l) => l)
    .map((s) => s.split(' '))
    .map((s) => ({ d: s[0], s: Number(s[1]) }));
}

function Rope(size) {
  this.trace = new Set();
  const knots = Array(size)
    .fill(0)
    .map((a) => ({ x: 0, y: 0 }));
  const head = knots[0];
  const tail = knots[size - 1];

  const shouldMove = (head, tail) => {
    if (tail.x >= head.x - 1 && tail.x <= head.x + 1 && tail.y >= head.y - 1 && tail.y <= head.y + 1) return false;
    return true;
  };

  const moveOneStep = () => {
    knots.reduce((head, tail) => {
      if (shouldMove(head, tail)) {
        // move every knots
        const dx = head.x - tail.x;
        const dy = head.y - tail.y;
        tail.x += dx > 0 ? 1 : dx < 0 ? -1 : 0;
        tail.y += dy > 0 ? 1 : dy < 0 ? -1 : 0;
      }
      return tail;
    });

    const { x, y } = tail;
    this.trace.add(`${x}:${y}`);
  };

  this.moveRight = (steps) => {
    while (steps > 0) {
      head.x += 1;
      moveOneStep();
      steps -= 1;
    }
  };

  this.moveLeft = (steps) => {
    while (steps > 0) {
      head.x -= 1;
      moveOneStep();
      steps -= 1;
    }
  };

  this.moveUp = (steps) => {
    while (steps > 0) {
      head.y += 1;
      moveOneStep();
      steps -= 1;
    }
  };

  this.moveDown = (steps) => {
    while (steps > 0) {
      head.y -= 1;
      moveOneStep();
      steps -= 1;
    }
  };
}

function solve(input, size) {
  const rope = new Rope(size);

  input.forEach((move) => {
    switch (move.d) {
      case 'R':
        rope.moveRight(move.s);
        break;
      case 'L':
        rope.moveLeft(move.s);
        break;
      case 'U':
        rope.moveUp(move.s);
        break;
      case 'D':
        rope.moveDown(move.s);
        break;
    }
  });
  return rope.trace.size;
}

//const input = readInput(`d${DAY}-sample1.txt`);
//const input = readInput(`d${DAY}-sample2.txt`);
const input = readInput(`d${DAY}-input.txt`);
console.log(`Part 1: ${solve(input, 2)}`);
console.log(`Part 2: ${solve(input, 10)}`);

File: 2022/D09/d09.py
"""
Advent Of Code 2022
Day 9: Rope Bridges part 1 & 2

https://adventofcode.com/2022/day/9
"""
from time import perf_counter as pfc
import os
import functools

YEAR = "2022"
DAY = "09"

def readInput(filename):
  with open(os.path.join(f"{YEAR}", f"D{DAY}", filename)) as f:
    lines = f.readlines()
  input = (d.strip('\n') for d in lines)
  return tuple(tuple((move[0], int(move[1]))) for move in (i.split(" ") for i in input))

class Rope:
  def __init__(self, size) -> None:
    self.knots = tuple([0, 0] for i in range(size))
    self.head = self.knots[0]
    self.tail = self.knots[size - 1]
    self.trace = set()

  def shouldMove(self, head, tail):
    if (tail[0] >= head[0] - 1 and tail[0] <= head[0] + 1 and tail[1] >= head[1] - 1 and tail[1] <= head[1] + 1):
      return False
    return True

  def moveOneStep(self):
    def _move(head, tail):
      if self.shouldMove(head, tail):
        dx = head[0] - tail[0]
        dy = head[1] - tail[1]
        tail[0] += 1 if dx > 0 else -1 if dx < 0 else 0
        tail[1] += 1 if dy > 0 else -1 if dy < 0 else 0
      return tail

    functools.reduce(_move, self.knots)
    self.trace.add(tuple(self.tail))

  def moveRight(self, steps):
    while (steps > 0):
      self.head[0] += 1
      self.moveOneStep()
      steps -= 1

  def moveLeft(self, steps):
    while (steps > 0):
      self.head[0] -= 1
      self.moveOneStep()
      steps -= 1

  def moveUp(self, steps):
    while (steps > 0):
      self.head[1] += 1
      self.moveOneStep()
      steps -= 1

  def moveDown(self, steps):
    while (steps > 0):
      self.head[1] -= 1
      self.moveOneStep()
      steps -= 1

def solve(input, size):
  startTime = pfc()
  rope = Rope(size)

  for move in input:
    d, s = move
    if (d == "R"):
      rope.moveRight(s)
    elif (d == "L"):
      rope.moveLeft(s)
    elif (d == "U"):
      rope.moveUp(s)
    elif (d == "D"):
      rope.moveDown(s)

  print(f"Time: {pfc() - startTime:.4}")
  return len(rope.trace)

input = readInput(f"d{DAY}-input.txt")
print(f"Part 1 for puzzle input: {solve(input,2)}")
print(f"Part 2 for puzzle input: {solve(input,10)}")
